File: api/schemas/brain.py
from __future__ import annotations

from typing import Any, Optional

from pydantic import BaseModel


class UploadResponse(BaseModel):
    file_id: str
    filename: str
    size: int


class AnalyzeRequest(BaseModel):
    file_id: str
    patient_case_id: str
    age: float = 72.0
    sex: str = "F"


class CNNResult(BaseModel):
    prediction: str
    confidence: float
    prob_cn: float
    prob_ad: float
    clinical_age: Optional[float] = None
    clinical_sex: Optional[str] = None


class AnalyzeResult(BaseModel):
    analysis_id: str
    result: CNNResult
    report_paths: dict  # {md: str, html: str, pdf: Optional[str]}


class GradCAMRequest(BaseModel):
    file_id: str
    patient_case_id: str
    analysis_id: str
    orientation: str = "multi"
    display_mode: str = "overlay"
    age: float = 72.0
    sex: str = "F"


class GradCAMImage(BaseModel):
    image_path: str   # relative path from project root
    image_url: str    # /files/... URL
    caption: str
    orientation: str


class GradCAMResult(BaseModel):
    images: list[GradCAMImage]


class ExplainRequest(BaseModel):
    patient_case_id: str
    analysis_id: str
    result: dict
    gradcam_info: Optional[dict] = None
    question: Optional[str] = None


class ExplainResult(BaseModel):
    answer: str
    sources: list[dict]


class ChatRequest(BaseModel):
    message: str
    cnn_result: dict
    gradcam_info: Optional[dict] = None
    history: list[dict] = []

File: frontend/src/api/brain.ts
import client from './client'
import type { ChatMessage } from '../types'

export const brainApi = {
  upload: async (file: File) => {
    const form = new FormData()
    form.append('file', file)
    const { data } = await client.post<{ file_id: string; filename: string; size: number }>(
      '/api/brain/upload',
      form
    )
    return data
  },

  analyze: async (file_id: string, patient_case_id: string, age: number, sex: string) => {
    const { data } = await client.post<{ task_id: string }>('/api/brain/analyze', {
      file_id,
      patient_case_id,
      age,
      sex,
    })
    return data
  },

  gradcam: async (
    file_id: string,
    patient_case_id: string,
    analysis_id: string,
    orientation: string,
    display_mode: string,
    age: number,
    sex: string
  ) => {
    const { data } = await client.post<{ task_id: string }>('/api/brain/gradcam', {
      file_id,
      patient_case_id,
      analysis_id,
      orientation,
      display_mode,
      age,
      sex,
    })
    return data
  },

  explain: async (
    patient_case_id: string,
    analysis_id: string,
    result: object,
    gradcam_info?: object,
    question?: string
  ) => {
    const { data } = await client.post<{ task_id: string }>('/api/brain/explain', {
      patient_case_id,
      analysis_id,
      result,
      gradcam_info,
      question,
    })
    return data
  },

  chat: async (
    message: string,
    cnn_result: object,
    gradcam_info?: object,
    history: ChatMessage[] = []
  ) => {
    const { data } = await client.post<{ answer: string; sources: unknown[] }>('/api/brain/chat', {
      message,
      cnn_result,
      gradcam_info,
      history,
    })
    return data
  },
}
